packages/jupyter-ai/jupyter_ai/personas/persona_manager.py

```diff
@@ -281,22 +285,69 @@ def route_message(self, new_message: Message):
         Method that routes an incoming message to the correct persona by calling
         its `process_message()` method.
 
-        - (TODO) If the chat contains only one persona & one user, then this
-        method routes all new messages to that persona.
+        - If the chat has more than one user, then this requires
+          `@`- mention to route the message to the persona
+
+        - If the chat contains only one persona & one user, then this
+          method routes all new messages to that persona.
+
+        - If the message contains `@`-mentioned personas, it routes to those personas
+          and updates the last_mentioned_persona.
+
+        - If no personas are mentioned and there is a last_mentioned_persona, it routes
+          to that persona.
 
-        - Otherwise, this method only routes new messages to personas that are
-        `@`-mentioned in the message.
+        - Otherwise, it does not route the message to any persona.
         """
+
+        # Gather routing context
+        human_users = self.get_active_human_users()
+        sender_is_persona = is_persona(new_message.sender)
         mentioned_personas = self.get_mentioned_personas(new_message)
-        if not len(mentioned_personas):
+
+        human_user_count = len(human_users)
+        persona_count = len(self.personas)
+        mentioned_count = len(mentioned_personas)
+
+        # Multi-user chat: require explicit @-mentions only
+        if human_user_count > 1:
+            for persona in mentioned_personas:
+                self.event_loop.create_task(persona.process_message(new_message))
             return
 
-        mentioned_persona_names = [persona.name for persona in mentioned_personas]
-        self.log.info(
-            f"Received new user message mentioning the following personas: {mentioned_persona_names}."
-        )
-        for persona in mentioned_personas:
-            self.event_loop.create_task(persona.process_message(new_message))
+        # Don't route persona replies without mentions
+        if sender_is_persona and mentioned_count == 0:
+            return
```

Comment:
Can this be moved above the `if human_user_count > 1` check? Seems safer to enforce this constraint first to prevent persona reply loops.

Reply:
Done!